File: day-8/python/main.py
import sys
import numpy as np

def parse_input():
    if len(sys.argv) != 2:
        sys.exit('I need a file')
    data = []
    with open(sys.argv[1]) as fd:
        for line in fd:
            row = [int(t) for t in line.strip()]
            data.append(row)
    return np.array(data)


def nearest_obstacle(block, v):
    """ Check if the array block "hides" v
    """
    i = 0
    hides = False
    for obstacle in block:
        i += 1
        if v <= obstacle:
            hides = True
            break
    return i, hides


def candidate_ranges(data, row, column, n):
    return [[range(row - 1, -1, -1), column],  # bottom
            [range(row + 1, n), column], # top
            [row, range(column - 1, -1, -1)], # left
            [row, range(column + 1, n)]] # right


def check_visibility(data, row, column, n):
    # boundary nodes
    if row == 0 or column == 0 or row == n - 1 or column == n - 1:
        return True
    # inner nodes
    v = data[row, column]
    r = True
    for cr in candidate_ranges(data, row, column, n):
        if not nearest_obstacle(data[cr[0], cr[1]], v)[1]:
            return True
    return False


def problem_1(data):
    n = len(data)
    count = 0
    for i in range(n):
        for j in range(n):
            if check_visibility(data, i, j, n):
                count += 1
    return count


def count_neighbors(data, row, column, n):
    v = data[row, column]
    r = 1
    for cr in candidate_ranges(data, row, column, n):
         r *= nearest_obstacle(data[cr[0], cr[1]], v)[0]
    return r


def problem_2(data):
    n = len(data)
    r = 0
    for i in range(n):
        for j in range(n):
            r = max(r, count_neighbors(data, i, j, n))
    return r


def main():
    data = parse_input()
    print("Day 8, problem 1: ", problem_1(data))
    print("Day 8, problem 2: ", problem_2(data))


if __name__ == '__main__':
    main()

File: day-8/rust/src/main.rs
use std::env;
use std::fs;
use std::process;
use itertools::Itertools;

fn parse_args() -> String {
    let mut args = env::args();
    if args.len() != 2 {
        eprintln!("Invalid arguments: I need the input file and nothing else");
        process::exit(1);
    }
    args.nth(1).expect("Already checked")
}

fn read_grid(filename: &String) -> (Vec<u32>, usize)  {
    let data = fs::read_to_string(filename).expect("Failed to read input file");
    let raw: Vec<u32>
        = data.trim().chars().filter(|x| x != &'\n')
            .map(|x| x.to_digit(10).unwrap()).collect();
    let n = (raw.len() as f64).sqrt() as usize;
    if n * n != raw.len() {
        eprintln!("Invalid data dimensions");
        process::exit(1);
    }
    (raw, n)
}

fn check_visibility(grid: &Vec<&[u32]>,
                    row: usize,
                    column: usize,
                    n: usize) -> bool {
    if row == 0 || column == 0 || row == n - 1 || column == n - 1 {
        return true;
    }
    let v = grid[row][column];
    let h1 = grid[0..row].iter()
        .map(|x| x[column])
        .filter(|x| x >= &v).count() == 0;
    let h2 = grid[row + 1..n].iter().map(|x| x[column])
        .filter(|x| x >= &v).count() == 0;
    let v1 = grid[row][0..column].iter().filter(|x| x >= &&v).count() == 0;
    let v2 = grid[row][column + 1..n].iter().filter(|x| x >= &&v).count() == 0;

    h1 || h2 || v1 || v2
}

fn problem_1(data: &Vec<u32>, n: usize) -> u32 {
    let grid: Vec<&[u32]> = data.chunks(n).collect();
    let mut count = 0;
    for i in 0..n {
        for j in 0..n {
            if check_visibility(&grid, i, j, n) {
                count += 1;
            }
        }
    }
    count
}

fn obstacle_distance(grid: &Vec<&[u32]>,
                     row: usize,
                     column: usize,
                     n: usize) -> u32 {
    let v = grid[row][column];

    let h1 = if let Some(t) = grid[0..row].iter().map(|x| x[column]).rev()
        .find_position(|x| x >= &v) { t.0 + 1 } else { row };
    let h2 = if let Some(t) = grid[row + 1..n].iter().map(|x| x[column])
        .find_position(|x| x >= &v) { t.0 + 1 } else { n - row - 1 };
    let v1 = if let Some(t) = grid[row][0..column].iter().rev()
        .find_position(|x| *x >= &v) { t.0 + 1 } else { column };
    let v2 = if let Some(t) = grid[row][(column + 1)..n].iter()
        .find_position(|x| *x >= &v) { t.0 + 1 } else { n - column - 1};

    (h1 * h2 * v1 * v2).try_into().unwrap()
}

fn problem_2(data: &Vec<u32>, n: usize) -> u32 {
    let grid: Vec<&[u32]> = data.chunks(n).collect();
    let mut r = 0;
    for i in 0..n {
        for j in 0..n {
            if check_visibility(&grid, i, j, n) {
                r = std::cmp::max(r, obstacle_distance(&grid, i, j, n))
            }
        }
    }
    r
}

fn main() {
    let filename = parse_args();
    let (data, n) = read_grid(&filename);
    println!("Day 8, problem 1: {}", problem_1(&data, n));
    println!("Day 8, problem 2: {}", problem_2(&data, n));
}
